container_images/pytest/main.py

```diff
@@ -0,0 +1,173 @@
+  "--override-ini=junit_family=xunit1",
+  # Within the junit report, add a namespace for the environment,
+  # to ensure that tests from different interpreters aren't
+  # considered duplicates.
+  "--junit-prefix={{envname}}",
+  # Write a new junit xml file for each interpreter. The filename pattern
+  # is defined in our Prow instance's configuration:
+  #  https://github.com/GoogleCloudPlatform/oss-test-infra/blob/cc1f0cf1ffecc0e3b75664b0d264abc6165276d1/prow/oss/config.yaml#L98
+  "--junit-xml={artifact_dir}/junit-{{envname}}.xml",
+]
+
+
+def setup_execution_root(package_root: str) -> str:
+  """Create execution directory and copy project's code to it.
+
+  Tox creates in-directory files, and its execution fails when the
+  source tree is mounted read-only.
+
+  Args:
+    package_root: Absolute path to the Python package's root directory.
+
+  Returns:
+    Absolute path to execution root.
+  """
+  assert (os.path.isabs(package_root) and
+          os.path.exists(package_root))
+
+  exec_root = tempfile.mkdtemp()
+  shutil.copytree(package_root, exec_root, dirs_exist_ok=True)
+  return exec_root
+
+
+def to_tox_version(py_version: str):
+  """Convert `{major}.{minor}` to `py{major}{minor}`"""
+  if re.fullmatch(r"\d.\d{1,2}", py_version):
+    major, minor = py_version.split(".")
+    return "py{major}{minor}".format(major=major, minor=minor)
+  else:
+    sys.exit("Invalid version number: " + py_version)
+
+
+def write_tox_ini(artifact_dir):
+  """Read project configuration from pyproject.toml, and write a new tox.ini
+  file.
+
+  The tox.ini file is generated to ensure that tests are run consistently,
+  and that test reports are written to the correct location.
+  """
+  assert os.path.exists("pyproject.toml")
+
+  with open("pyproject.toml") as f:
+    project_cfg = toml.load(f)
+
+  # Read the [tool.gcp-guest-pytest] section of pyproject.toml.
+  test_cfg = project_cfg.get("tool", {}).get("gcp-guest-pytest", {})
+
+  # Which interpreters to enable.
+  envlist = [to_tox_version(env) for env in test_cfg.get("envlist", [])]
+
+  # Test-specific dependencies to install.
+  test_deps = test_cfg.get("test-deps", [])
+
+  if not envlist:
+    sys.exit("pyproject.toml must contain a section [tool.gcp-guest-pytest] "
+             'with a key "envlist" and at least one interpreter.')
+
+  config = configparser.ConfigParser()
+  config["tox"] = {
+    "envlist": ", ".join(envlist),
+  }
+
+  config["testenv"] = {
+    "envlogdir": "{artifact_dir}/tox/{{envname}}"
+      .format(artifact_dir=artifact_dir),
+    "deps": "\n\t".join(_common_test_deps + test_deps),
+    "commands": " ".join(_per_interpreter_command)
+      .format(artifact_dir=artifact_dir),
+  }
+
+  if os.path.exists("tox.ini"):
+    print("Removing existing tox.ini file.")
+    os.remove("tox.ini")
+
+  with open("tox.ini", "w") as f:
+    config.write(f)
+
+
+def archive_configs(dst: str):
```

Comment:
what happens if dst exists, but is a file?

Reply:
It wouldn't behave as expected! Added an assertion that dst is a directory if it exists.